src/libopensc/card.c

```diff
@@ -1216,13 +1225,13 @@ sc_algorithm_info_t *sc_card_find_alg(sc_card_t *card,
 	for (i = 0; i < card->algorithm_count; i++) {
 		sc_algorithm_info_t *info = &card->algorithms[i];
 
-		if (info->algorithm != algorithm)
-			continue;
 		if (param && (info->algorithm == SC_ALGORITHM_EC ||
 			info->algorithm == SC_ALGORITHM_EDDSA ||
 			info->algorithm == SC_ALGORITHM_XEDDSA)) {
 			if (sc_compare_oid((struct sc_object_id *)param, &info->u._ec.params.id))
 				return info;
+		} else if (info->algorithm != algorithm) {
+			continue;
```

Comment:
In pervious code, SC_ALGORITHM_EDDSA only referred to ED25519, But in this PR it can refer to ED25519 or ED448 (Same for the X curves). to make sure the OID must also be checked. The OID implies the key length.

7ae749e135 from 24-05-16, part of this PR  says: "card_find_alg will first check if  info->algorithm is one that uses sc_ec_parameters and then checks that the OIDs match. then check if keylength match."

fb3e7203b2d63b28 also from this PR, calls `sc_pkcs15_fix_ec_parameters` to make sure algorithms are set.    

I think I made that change based on some gdb testing, but now am not sure why. It may not be need.

Reply:
Note the "gdb" testing code was there because of #3000 which has bee withdrawn.

Reply:
Thanks for clarification. I think its reasonable.